File: test/corpus/control/with_result.exs
defmodule RustQ.Corpus.Control.WithResult do
  @moduledoc "Nested with lowering without synthetic if true blocks."

  use RustQ.Meta

  alias RustQ.Type, as: R

  @spec decode_a(term()) :: R.nif_result(R.u32())
  defrust decode_a(term) do
    case term do
      1 -> {:ok, 10}
      _ -> {:error, badarg()}
    end
  end

  @spec decode_b(term()) :: R.nif_result(R.u32())
  defrust decode_b(term) do
    case term do
      2 -> {:ok, 20}
      _ -> {:error, badarg()}
    end
  end

  @spec decode(term()) :: R.nif_result(R.u32())
  defrust decode(term) do
    with {:error, _a_reason} <- decode_a(term),
         {:error, _b_reason} <- decode_b(term) do
      {:error, badarg()}
    else
      {:ok, value} -> {:ok, value + 1}
      {:error, _reason} -> {:error, badarg()}
    end
  end
end

File: test/corpus/control/with_result.rs
fn decode_a<'a>(term: Term<'a>) -> NifResult<u32> {
    match term {
        1 => Ok(10),
        _ => Err(rustler::Error::BadArg),
    }
}

fn decode_b<'a>(term: Term<'a>) -> NifResult<u32> {
    match term {
        2 => Ok(20),
        _ => Err(rustler::Error::BadArg),
    }
}

fn decode<'a>(term: Term<'a>) -> NifResult<u32> {
    match decode_a(term) {
        Err(_a_reason) => match decode_b(term) {
            Err(_b_reason) => Err(rustler::Error::BadArg),
            __rustq_with_value => match __rustq_with_value {
                Ok(value) => Ok(value + 1),
                Err(_reason) => Err(rustler::Error::BadArg),
            },
        },
        __rustq_with_value => match __rustq_with_value {
            Ok(value) => Ok(value + 1),
            Err(_reason) => Err(rustler::Error::BadArg),
        },
    }
}
